### PR TITLE
[SPARK-34775][SQL] Push down limit through window when partitionSpec is not empty

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
This is a followup from #31691. Push down limit through Window when the partitionSpec of all window functions is not empty and the same order is used.
Push down limit through Window when the partitionSpec of all window functions is not empty

And the origin author is @leoluan2009, since he didn't reply for long and did this follow up after invitation 

### Why are the changes needed?
Improve query performance.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added UT

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
ping @wangyum

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/LimitPushDownThroughWindow.scala [hunk 1]
```diff
@@ -42,17 +42,19 @@ object LimitPushDownThroughWindow extends Rule[LogicalPlan] {
     _.containsAllPatterns(WINDOW, LIMIT), ruleId) {
     // Adding an extra Limit below WINDOW when the partitionSpec of all window functions is empty.
     case LocalLimit(limitExpr @ IntegerLiteral(limit),
-        window @ Window(windowExpressions, Nil, orderSpec, child))
+    window @ Window(windowExpressions, partitionSpec, orderSpec, child))
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/LimitPushDownThroughWindow.scala:45]: Nit: indentation.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/LimitPushdownThroughWindowSuite.scala [hunk 2]
```diff
@@ -155,36 +155,164 @@ class LimitPushdownThroughWindowSuite extends PlanTest {
       WithoutOptimize.execute(correctAnswer.analyze))
   }
 
-  test("Should not push down if partitionSpec is not empty") {
+  test("Should not push down when child's maxRows smaller than limit value") {
```
Comment (wangyum) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/LimitPushdownThroughWindowSuite.scala:158]: Can we do not move the order? To reduce the change.

Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/LimitPushdownThroughWindowSuite.scala]: > Can we do not move the order? To reduce the change.

Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/LimitPushDownThroughWindow.scala (hunk 1 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/LimitPushDownThroughWindow.scala]: > Nit: indentation.

Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/LimitPushDownThroughWindow.scala [hunk 3]
```diff
@@ -42,17 +42,19 @@ object LimitPushDownThroughWindow extends Rule[LogicalPlan] {
...
     // Adding an extra Limit below WINDOW when the partitionSpec of all window functions is empty.
     case LocalLimit(limitExpr @ IntegerLiteral(limit),
-        window @ Window(windowExpressions, Nil, orderSpec, child))
+        window @ Window(windowExpressions, partitionSpec, orderSpec, child))
       if supportsPushdownThroughWindow(windowExpressions) && child.maxRows.forall(_ > limit) &&
         limit < conf.topKSortFallbackThreshold =>
       // Sort is needed here because we need global sort.
-      window.copy(child = Limit(limitExpr, Sort(orderSpec, true, child)))
+      window.copy(child = Limit(limitExpr,
+        Sort(partitionSpec.map(SortOrder(_, Ascending)) ++ orderSpec, true, child)))
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/LimitPushDownThroughWindow.scala:50]: @cloud-fan Do we need to changed the sort order from `NULLS FIRST` to `NULLS LAST`? Impala have changed it.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/LimitPushDownThroughWindow.scala:50]: is `NULLS LAST` faster?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/LimitPushDownThroughWindow.scala [hunk 4]
```diff
@@ -42,17 +42,19 @@ object LimitPushDownThroughWindow extends Rule[LogicalPlan] {
     _.containsAllPatterns(WINDOW, LIMIT), ruleId) {
     // Adding an extra Limit below WINDOW when the partitionSpec of all window functions is empty.
     case LocalLimit(limitExpr @ IntegerLiteral(limit),
-        window @ Window(windowExpressions, Nil, orderSpec, child))
+        window @ Window(windowExpressions, partitionSpec, orderSpec, child))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/LimitPushDownThroughWindow.scala:45]: can you provide a bit more context? why can we pushdown limit through window when there are partition specs?

**Pull request closed without merging**